parsers/julia.c

```diff
@@ -0,0 +1,1331 @@
+/*
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   This module contains functions for generating tags for Julia files.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"	/* must always come first */
+
+#include <string.h>
+
+#include "keyword.h"
+#include "parse.h"
+#include "entry.h"
+#include "options.h"
+#include "read.h"
+#include "routines.h"
+#include "vstring.h"
+#include "xtag.h"
+
+/*
+*   MACROS
+*/
+#define MAX_STRING_LENGTH 256
+
+/*
+*   DATA DEFINITIONS
+*/
+typedef enum {
+    K_CONSTANT,
+    K_FUNCTION,
+    K_FIELD,
+    K_MACRO,
+    K_MODULE,
+    K_STRUCT,
+    K_TYPE,
+    K_IMPORT,
+    K_NONE
+} JuliaKind;
+
+static kindDefinition JuliaKinds [] = {
+    { true, 'c', "constant", "Constants"    },
+    { true, 'f', "function", "Functions"    },
+    { true, 'g', "field",    "Fields"       },
+    { true, 'm', "macro",    "Macros"       },
+    { true, 'n', "module",   "Modules"      },
+    { true, 's', "struct",   "Structures"   },
+    { true, 't', "type",     "Types"        },
+    { true, 'x', "unknown",  "Imported name"}
+};
```

Comment:
In u-ctags,  we use "reference tag" to tag for X in "import X" because
"import X" doesn't define X. X is defined somewhere. "import X" refers it.

See https://github.com/universal-ctags/ctags/issues/2428, 
https://docs.ctags.io/en/latest/man/ctags-lang-python.7.html, and parsrs/python.c.

As far as I can remember, I don't write any document the way to make reference tags. So whether I can ask you to use a reference tag for extracting X in "import X".  I can work on this topic instead of you.  Tell me if I should do (after merging your parser). In such a case, what I'm afraid of is that the impact of my change to Geany.
You are working on Julia's support in Geany. Introducing reference tags may break your work partially.
To avoid it, we have to cooperate.

Reply:
I don't understand how reference tags work. Can you explain a bit? I found it in python.c and thought it was appropriate for julia too. Just for you to know, in Julia, imports work like in python `from <module> import *` so I don't know if looking for definition in another file would work...

Reply:
The description of reference tags for users:
By default, ctags records only definition tags.
If you also want ctags to record reference tags, specify `--extras=+r`.
A reference tag has a kind like function, variable, etc. This is the same as definition tags.
In addition, a tag has roles. A role is a subcategory of a kind. The roles are not shown by
default. specify `--fileds=+r`.

```
x = 1
y++
def z
```
In this example 
```
x ... kind:variable roles:assigned
y ... kind:variable roles:assigned,incremented
z ... kind:varaible roles:def
```
`def` in roles means z is tagged as a definition tag.
x and y have no `def' in `roles` field.  Therefore, they are tagged as reference tags.

What kind of roles other than `def` is defined can be listed with `--list-roles` option.
```
$ ./ctags --list-roles=Python
#KIND(L/N) NAME               ENABLED DESCRIPTION
i/module   imported           on      imported modules
i/module   indirectlyImported on      module imported in alternative name
i/module   namespace          on      namespace from where classes/variables/functions are imported
x/unknown  imported           on      imported from the other module
x/unknown  indirectlyImported on      classes/variables/functions/modules imported in alternative name
$ ./ctags --list-roles=Python.'{module}'
#KIND(L/N) NAME               ENABLED DESCRIPTION
i/module   imported           on      imported modules
i/module   indirectlyImported on      module imported in alternative name
i/module   namespace          on      namespace from where classes/variables/functions are imported

```
CURRENTLY, Reference tags are not popular. 
However,  in a limited area, I'm pushing introducing reference tags. See #2428.

The description of reference tags for parser developers.
Follow the steps to introduce a reference tag to your parser.

1. define a role in a kind like:
```
typedef enum {
	...
	K_MODULE,
	...
	COUNT_KIND
} pythonKind;

typedef enum {
	PYTHON_MODULE_IMPORTED,
	PYTHON_MODULE_NAMESPACE,
	PYTHON_MODULE_INDIRECTLY_IMPORTED,
} pythonModuleRole;

static roleDefinition PythonModuleRoles [] = {
	{ true, "imported",
	  "imported modules" },
	{ true, "namespace",
	  "namespace from where classes/variables/functions are imported" },
	{ true, "indirectlyImported",
	  "module imported in alternative name" },
};

/* if a kind is never used for making a definition tag,
   set `referenceOnly` to true. */
static kindDefinition PythonKinds[COUNT_KIND] = {
	...
	{true, 'i', "module",    "modules",
	 .referenceOnly = true,  ATTACH_ROLES(PythonModuleRoles)},
        ...
};
```
2. make a referenec tag like
```
		tagEntryInfo e; 
		int kindIndex = K_MODULE;
		int roleIndex = PYTHON_MODULE_NAMESPACE;
		initRefTagEntry (&e, "tagNameInCString",
		                 kindIndex, roleIndex);
		...

		markTagExtraBit (&e, XTAG_REFERENCE_TAGS);

		return makeTagEntry (&e);
```
or
```
		makeSimpleRefTag (tagNameinVString, K_MODULE, PYTHON_MODULE_NAMESPACE);
```